src/Database/LSMTree/Internal/Run.hs

```diff
@@ -141,3 +152,65 @@ fromWriteBuffer fs fsPaths buffer = do
       addFullKOp fs mrun k e
     fromMutable fs mrun
+
+-- | Load a previously written run from disk, checking each file's checksum
+-- against the checksum file.
+--
+-- TODO: define and throw exceptions instead of using 'error'?
+openFromDisk :: HasFS IO h -> RunFsPaths -> IO (Run (FS.Handle h))
+openFromDisk fs lsmRunFsPaths = do
+    -- verify checksums
+    expectedChecksums <-
+      fmap (fromMaybe (error "invalid checksums file") . fromChecksumsFile) $
+        CRC.readChecksumsFile fs (runChecksumsPath lsmRunFsPaths)
+    sequence_ (checkCRC <$> expectedChecksums <*> runFsPaths lsmRunFsPaths)
+
+    let paths = runFsPaths lsmRunFsPaths
+    checkCRC (forRunKOps expectedChecksums) (forRunKOps paths)
+    checkCRC (forRunBlob expectedChecksums) (forRunBlob paths)
+    bsFilter <- readCRC (forRunFilter expectedChecksums) (forRunFilter paths)
+    bsIndex <- readCRC (forRunIndex expectedChecksums) (forRunIndex paths)
+
+    -- try parsing files
+    let lsmRunFilter =
+          either (error . ("invalid filter file: " <>)) id $
+            bloomFilterFromSBS bsFilter
+    let (lsmRunNumEntries, lsmRunIndex) =
+          either (error . ("invalid index file: " <>)) id $
+            Index.fromSBS bsIndex
+
+    lsmRunKOpsFile <- FS.hOpen fs (runKOpsPath lsmRunFsPaths) FS.ReadMode
+    lsmRunBlobFile <- FS.hOpen fs (runBlobPath lsmRunFsPaths) FS.ReadMode
+    lsmRunRefCount <- newIORef 1
+    return Run {..}
+  where
+    checkCRC :: CRC.CRC32C -> FS.FsPath -> IO ()
+    checkCRC expected fp = do
+      checksum <- CRC.readFileCRC32C fs fp
+      when (expected /= checksum) $
+        error $ "invalid checksum for file " <> show fp
+
+    -- This function should almost always directly read the whole file into a
+    -- contiguous ShortByteString/ByteArray without additional copies.
+    -- To be really sure about it, we need fs-api support for directly reading
+    -- into a buffer, which is currently work in progress:
+    -- https://github.com/input-output-hk/fs-sim/pull/46
+    readCRC :: CRC.CRC32C -> FS.FsPath -> IO SBS.ShortByteString
+    readCRC expected fp = FS.withFile fs fp FS.ReadMode $ \h -> do
+        -- Try to read the whole file into a single strict ByteString,
+        -- which should almost always succeed. In this case, 'mconcat' is O(1).
+        size <- FS.hGetSize fs h
+        bs <- mconcat <$> go h size CRC.initialCRC32C
+        case tryCheapToShort bs of
+          Right sbs -> return sbs
+          -- Should not happen, so for now let's make sure it doesn't.
+          -- TODO: fall back to BS.toShort?
+          Left err  -> error $ "tryCheapToShort: " <> err
+      where
+        go _ 0 crc
+          | crc == expected = return []
+          | otherwise       =  error $ "invalid checksum for file " <> show fp
+        go h n crc = do
+            (bs, crc') <- CRC.hGetSomeCRC32C fs h n crc
+            -- Keep reading, in case there is more.
+            (bs :) <$> go h (n - fromIntegral (BS.length bs)) crc'
```

Comment:
The reading part can be a proper abstraction in the style of [`hGetAll`](https://input-output-hk.github.io/fs-sim/fs-api/System-FS-API-Lazy.html#v:hGetAll) and `hGetExactly`. That function (`hGetExactlyCRC32C`) can live in the `CRC32C` module, and it can return a lazy bytestring instead of a list strict bytestrings

Reply:
I implemented it as a thin wrapper around `hGetExactly`. I think doing it this way (vs. interleaving disk I/O and CRC checks) should be fine, since the number of chunks should be very small.

Reply:
> I implemented it as a thin wrapper around `hGetExactly`. I think doing it this way (vs. interleaving disk I/O and CRC checks) should be fine, since the number of chunks should be very small.

:+1:  It's also not on a performance-critical path. At first glance, I think interleaving would also be more tricky when we start using `HasBufFS`